### ISSUE TITLE
Need OneTimeTearDown call once for multiAssembly run in parallel

### ISSUE THREAD

**Issue by shmin2** [nunit/nunit#3149]
Need "trigger" when ALL tests finished. [OneTimeTearDown] don't help with it.
Have NUnit3 any tool that notify run finish?

My solution:
A project:
   -- TestBaseClass
B project:
  -- Tests1.cs : TestBaseClass
C project:
  -- Tests2.cs : TestBaseClass
....and more

I add to all projects [assembly: Parallelizable(ParallelScope.All)] [assembly: LevelOfParallelism(2)].
A.TestBaseClass have method [OneTimeTearDown], but it calls every time when assembly run finished, but expected only once.

Environment:
VS 2017
All projects .netCore 2.1
NUnit 3.11.0
NUnit3TestAdapter 3.12.0

**Comment by CharliePoole** [comment]
First of all, I can understand why you may want this. It makes sense.

The nunit/nunit project can't give it to you, however, because this project is about the nunit framework, which only deals with one assembly at a time. Hypothetically, we could give you a way to mark an assembly as a "setup assembly" and allow either the engine or various runners to identify those assemblies and run them before and after everything else, but the main work would either be in the engine itself or the runner.

Of the two, the engine makes most sense on the surface. If we implement it in the engine then it's there for all runners to use. But there is a big "however."

You are using the NUnit3 TestAdapter, which does __not__ run assemblies in parallel. It always runs them separately in sequence. I suspect that any parallelism you are getting is from the dotnet CLI, not from NUnit itself. Each time the NUnit adapter is invoked, it is called for one assembly and doesn't know anything about other assemblies.

Bottom line, I think this issue should be moved to either the nunit-console repo (if we want the engine to handle it) or the nunit3-vs-adapter repo (if we want the adapter to handle it) or (possibly) just closed. I'll leave that for one of the project leads to decide.

**Comment by rprouse** [comment]
May I ask what types of actions you would like to perform in an overall setup or teardown? Your request is a logical extension of what is available, but I'm not sure how widely used it would be.

I think we can leave this issue here for now. Even though the execution will need to be handled in the Engine, we will still need an attribute to signal the one time setup/teardown assembly.

**Comment by shmin2** [comment]
Before test run i generate (very very slow operation) X users and create shared pool between tests. Between assemblies i move them through JSON file saved on hard drive. After ALL tests finished and released into pool their users, i want to start delete them. 

Now I do this with additional Task/Job on CI server but if somebody "skip/lost" this step test account will have "dirty" users. And when somebody run tests locally from Visual Studio they can forgot to clean.

So idea is to call Clean after all tests finished automatically whatever start local or CI. And run can be parallel.

**Comment by rprouse** [comment]
Thanks, I understand your use case. Personally, I find it to be an anti-pattern to have dependencies in your tests that span multiple test assemblies. That said, I can fully understand why you are doing it the way you are and can sympathize.

**Comment by vanderkyl** [comment]
I have a similar problem. In my case, we are using the OneTimeSetUp and OneTimeTearDown methods to send updates to a web application that is used as a test manager GUI.

This means that when a group of tests coming from multiple assemblies is launched, our UI is receiving more than one update that the tests have started or finished. The number of updates obviously matches the number of assemblies in the test run.

Another related problem we are having is that when using the --workers console option, it is applied to each assembly instead of all of the tests. For instance if I set the workers value to 4, and run a group of 8 tests (4 in one assembly and 4 in another), all 8 tests will start running instead of just 4 at a time.

Any feedback is welcome. Thank you!

**Comment by shmin2** [comment]
To solve --workers problem I use "dotnet test **/m:1**". M argument goes from MSBuild [https://docs.microsoft.com/ru-ru/visualstudio/msbuild/building-multiple-projects-in-parallel-with-msbuild?view=vs-2019](MSBuild docs). 

After add this argument only 1 assembly will be run but all workers will parallel run tests.

**Comment by rprouse** [comment]
This issue is marked as an idea but hasn't had contributions in years so I am closing. If anyone is interested in working on this idea, post your interest and the team will consider reopening.

**Comment by schmitch** [comment]
@rprouse I actually looked into something like that at the moment.
I quite liked your idea about a "setup assembly".

For myself the goal would be to have a way to build a docker container for `https://testcontainers.com/` basically my company has different tests that still test the same database and everytime a new assembly gets started it will build it's own container which will contain **all** migrations **applied**.

however our migrations are quite big already and it takes like 10 seconds to apply them, thus its way slower than to build the container upfront and run the test with them. starting the container per test is than ultra fast and our tests can change the database however they like and we have no downsides. which we have with all other solutions like respawn, etc.
